text/3729-sized-hierarchy.md

```diff
@@ -0,0 +1,1412 @@
+│ │ │ ┃ ┃ const Sized       ┃ ┃ Sized                            │ │ │
+│ │ │ ┃ ┃ {type, target}    ┃ ┃ {type, target, runtime env}      │ │ │
+│ │ │ ┃ ┗━━━━━━━━━━━━━━━━━━━┛ ┃                                  │ │ │
+│ │ └─╂───────────────────────╂──────────────────────────────────┘ │ │
+│ │   ┃                       ┃                                    │ │
+│ │   ┃ const ValueSized      ┃ ValueSized                         │ │
+│ │   ┃ {type, target, value} ┃ {type, target, runtime env, value} │ │
+│ │   ┗━━━━━━━━━━━━━━━━━━━━━━━┛                                    │ │
+│ └────────────────────────────────────────────────────────────────┘ │
+│  Pointee                                                           │
+│  {*}                                                               │
+└────────────────────────────────────────────────────────────────────┘
+```
+
+`const Sized` is implemented on types which require knowledge of only the
+type and target platform in order to compute their size. For example, `usize`
+implements `const Sized` as knowing only the type is `usize` and the target is
+`aarch64-unknown-linux-gnu` then we can know the size is eight bytes, and
+likewise with `armv7-unknown-linux-gnueabi` and a size of four bytes.
+
+Some types' size can be computed using only knowledge of the type and target
+platform, but only at runtime, and these types implement `Sized` (notably,
+not `const`). For example, `svint8_t` is a scalable vector type (from
+[rfc#3268][rfc_scalable_vectors]) whose length depends on the specific
+implementation of the target (i.e. it may be 128 bits on one processor and
+256 bits on another).
+
+`const ValueSized` requires more knowledge than `const Sized` to compute the size:
+it may additionally require a value  (therefore `size_of` is not implemented for
+`ValueSized`, only `size_of_val`). For example, `[usize]` implements
+`const ValueSized` as knowing the type and target is not sufficient, the number of
+elements in the slice must also be known, which requires having the value.
+
+As `Sized` is to `const Sized`, `ValueSized` is to `const ValueSized`: `ValueSized`
+requires a value, knowledge of the type and target platform, and can only be
+computed at runtime. For example, `[svint8_t]` requires a value to know how
+many elements there are, and then information from the runtime environment
+to know the size of a `svint8_t`.
+
+`Pointee` is implemented by any type that can be used behind a pointer, which is
+to say, every type (put otherwise, these types may or may not be sized at all).
+For example, `Pointee` is therefore implemented on a `u32` which is trivially
+sized, a `[usize]` which is dynamically sized, a `svint8_t` which is
+runtime-sized and an `extern type` (from [rfcs#1861][rfc_extern_types]) which has
+no known size.
+
+All type parameters have an implicit bound of `const Sized` which will be
+automatically removed if a `Sized`, `const ValueSized`, `ValueSized` or
```

Comment:
presumably plus `~const Sized`, `~const ValueSized`?

What happens on `?ValueSized` and `?Pointee`? Remember that `?Trait` is currently permitted for all traits `Trait` (and leads to a non-lint warning if `Trait` ≠ `Sized`)

<sub>if this paragraph were to get updated, the corresponding one in section *`Sized` bounds* should be updated accordingly</sub>

Reply:
Fixed! I've noted that `?ValueSized` and `?Pointee` would be ignored and warn like other traits.